pygsp/graphs/fourier.py

```diff
@@ -88,26 +102,38 @@ def compute_fourier_basis(self, recompute=False):
         True
 
         """
+        if n_eigenvectors is None:
+            n_eigenvectors = self.N
 
-        if hasattr(self, '_e') and hasattr(self, '_U') and not recompute:
+        if (hasattr(self, '_e') and hasattr(self, '_U') and not recompute
+                and n_eigenvectors <= len(self.e)):
             return
 
         assert self.L.shape == (self.N, self.N)
-        if self.N > 3000:
-            self.logger.warning('Computing the full eigendecomposition of a '
-                                'large matrix ({0} x {0}) may take some '
-                                'time.'.format(self.N))
-
-        # TODO: handle non-symmetric Laplatians. Test lap_type?
-
-        self._e, self._U = np.linalg.eigh(self.L.toarray())
-        # Columns are eigenvectors. Sorted in ascending eigenvalue order.
-
-        # Smallest eigenvalue should be zero: correct numerical errors.
-        # Eigensolver might sometimes return small negative values, which
-        # filter's implementations may not anticipate. Better for plotting too.
-        assert -1e-12 < self._e[0] < 1e-12
-        self._e[0] = 0
+        if self.N**2 * n_eigenvectors > 3000**3:
+            self.logger.warning(
+                'Computing the {0} eigendecomposition of a large matrix ({1} x'
+                ' {1}) is expensive. Consider decreasing n_eigenvectors '
+                'or, if using the Fourier basis as a filter, using a '
+                'polynomial filter instead.'.format(
+                    'full' if n_eigenvectors == self.N else 'partial',
+                    self.N))
+
+        # TODO: handle non-symmetric Laplacians. Test lap_type?
+        if n_eigenvectors == self.N:
+            self._e, self._U = np.linalg.eigh(self.L.toarray())
+            # Columns are eigenvectors. Sorted in ascending eigenvalue order.
+
+            # Smallest eigenvalue should be zero: correct numerical errors.
+            # Eigensolver might sometimes return small negative values, which
+            # filter's implementations may not anticipate. Better for plotting
+            # too.
+            assert -1e-12 < self._e[0] < 1e-12
+            self._e[0] = 0
```

Comment:
Can we do those tests after the if-else branch?

Reply:
A partial eigendecomposition won't compute the smallest eigenvector, so there's no reason why the smallest eigenvector _would_ be 0.